Add support for zstd compression

LFrobeen opened the pull request:
This PR adds support for zstandard compression to the tarball layer implementation. It adds the application/vnd.oci.image.layer.v1.tar+zstd media type, a LayerCompression type and some extra logic in the tarball package. Moreover, this PR modifies [pkg/v1/partial/compressed.go](https://github.com/northflank/go-containerregistry/pull/2/files#diff-6632d2ac8e3d36f5972180430ed71640533372e8e1e3da9b377e8a5e7641c5af) so that it is able to unpack zstd compressed layers.

When creating a tarball layer, one can specify the compression algorithm using WithCompression and the compression level with WithCompressionLevel. By default it will still use gzip compression.
The changes should be fully backwards compatible.

imjasonh commented on pkg/v1/types/types.go:
```diff
@@ -24,7 +24,9 @@ const (
 	OCILayer                       MediaType = "application/vnd.oci.image.layer.v1.tar+gzip"
+	OCILayerZStd                   MediaType = "application/vnd.oci.image.layer.v1.tar+zstd"
 	OCIRestrictedLayer             MediaType = "application/vnd.oci.image.layer.nondistributable.v1.tar+gzip"
+	OCIRestrictedLayerZStd         MediaType = "application/vnd.oci.image.layer.nondistributable.v1.tar+zstd"
```
I'd prefer not to add this for now until we have a use case for it.

We might be able to get rid of non-distributable layers entirely: https://github.com/opencontainers/image-spec/pull/965

LFrobeen replied:
Ok, that makes sense. I'll remove it.

imjasonh reviewed (commented):
Wow, thanks for this! I've just taken a quick look so far but it looks great to me.

imjasonh commented:
It looks like there are some conflicts preventing CI from running. Can you rebase and we can see what the tests have to say?

LFrobeen commented:
Thanks for the feedback @imjasonh. I've removed the `OCIRestrictedLayerZStd` type and rebased the changes onto the main branch. 
So far I ran the unit tests locally and they appear to be successful. However, I should definitely add more tests for the new functionality. I was thinking about adding some more test cases to `tarbal/image_test.go` and `tarbal/layer_test.go`. Do you think that would be enough already, or are there other places where I should add more tests?

imjasonh commented:
> Thanks for the feedback @imjasonh. I've removed the `OCIRestrictedLayerZStd` type and rebased the changes onto the main branch. So far I ran the unit tests locally and they appear to be successful. However, I should definitely add more tests for the new functionality. I was thinking about adding some more test cases to `tarbal/image_test.go` and `tarbal/layer_test.go`. Do you think that would be enough already, or are there other places where I should add more tests?

That seems sufficient honestly. If this is something we expect to expose via a flag in `crane` it might be nice to have an e2e test for it, but I don't think we need to tackle that in this PR.

cc @jonjohnsonjr in case he has thoughts

jonjohnsonjr commented on pkg/v1/tarball/layer.go:
```diff
@@ -90,11 +102,26 @@ func (l *layer) MediaType() (types.MediaType, error) {
+		if compression == ZStd {
+			l.mediaType = types.OCILayerZStd
+		} else {
+			l.mediaType = types.DockerLayer
```
I would expect `WithCompression(None)` to set `l.mediaType` to `DockerUncompressedLayer` and produce uncompressed layers. WDYT?

I am also OK with punting on this for now, returning an error or documenting that None is invalid.

LFrobeen replied:
That makes sense, but I'm not quite sure how easy this would be to implement. My understanding is that the decision whether layers are produced with or without compression is made by consumers of the Layer interface when they access the data via `.Compressed()` or `.Uncompressed()`. Hence I'd have to modify all the code sites that use `.Compressed()` to make `WithCompression(None)` work, right?

I was thinking about alternatively just returning uncompressed data from the `.Compressed()` method when `WithCompression(None)` is set. But I felt that would conflict with semantics of the `.Compressed()` method which may be a bit confusing. Then again, this PR assumes the consumers of the `.Compressed()` method are agnostic to the compression format anyway, so one could argue that no compression can still be regarded "some kind of very inefficient compression format".

If you agree with this idea, I would implement it that way. Otherwise, I think the best option might be to drop `WithCompression(None)` altogether for now.

jonjohnsonjr replied:
> But I felt that would conflict with semantics of the .Compressed() method which may be a bit confusing.

Yeah that's fair. Having both the `Compressed` and `Uncompressed` methods is a design mistake. It would be better if we deprecated `Uncompressed` and renamed `Compressed`, but it's too late for that. I am fine with dropping this for now, but I think we want to return an error if we don't want to support it.

LFrobeen replied:
I modified the `WithCompression` method to ignore calls the `compression.None` argument and use `gzip` instead. Additionally it will print this warning:
```
Compression type 'none' is not supported for tarball layers; using gzip compression.
```

jonjohnsonjr commented on internal/zstd/zstd.go:
```diff
@@ -0,0 +1,114 @@
+	go func() error {
+		// TODO(go1.14): Just defer {pw,gw,r}.Close like you'd expect.
+		// Context: https://golang.org/issue/24283
+		gw, err := zstd.NewWriter(bw, zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(level)))
```
Nit: can these `gw` be `zw`?

jonjohnsonjr commented on internal/zstd/zstd.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// UnzipReadCloser reads compressed input data from the io.ReadCloser and
+// returns an io.ReadCloser from which uncompessed data may be read.
```
```suggestion
// returns an io.ReadCloser from which uncompressed data may be read.
```

jonjohnsonjr replied:
(And in the gzip version, old typo...)

LFrobeen replied:
Fixed.

jonjohnsonjr commented on internal/gzip/zip.go:
```diff
@@ -113,34 +113,5 @@ func Is(r io.Reader) (bool, error) {
-// If r implements Peek, we will use that directly, otherwise a small number
-// of bytes are buffered to Peek at the gzip header, and the returned
-// PeekReader can be used as a replacement for the consumed input io.Reader.
-func Peek(r io.Reader) (bool, PeekReader, error) {
```
Can you add a stub for `gzip.Peek` and `zstd.Peek`?

LFrobeen replied:
Sure, I'll just restore the original functionality, yes?

jonjohnsonjr commented on internal/zstd/zstd_test.go:
```diff
@@ -0,0 +1,101 @@
+		t.Error("ReadCloser: expected errRead, got", err)
+	}
+
+	//zr = ReadCloserLevel(io.NopCloser(strings.NewReader("zip me")), -10)
```
Need to clean this up?

LFrobeen replied:
Fixed.

jonjohnsonjr commented on pkg/v1/tarball/layer.go:
```diff
@@ -19,25 +19,37 @@ import (
+
+// The collection of known MediaType values.
+const (
+	None LayerCompression = "none"
```
I don't love these symbols being `tarball.None`, etc...

I am tempted to just expose a `compression` package so that we don't have to duplicate these everywhere... @imjasonh wdyt?

LFrobeen replied:
Yes I agree that `tarball.None` looks weird. I would also be in favor of creating a non-internal `compression` package.

LFrobeen replied:
I created a public `pkg/compression` package that contains the `Compression` enum.

jonjohnsonjr commented on pkg/v1/tarball/layer.go:
```diff
@@ -90,11 +102,26 @@ func (l *layer) MediaType() (types.MediaType, error) {
+func WithCompression(compression LayerCompression) LayerOption {
+	return func(l *layer) {
+		if compression == ZStd {
+			l.mediaType = types.OCILayerZStd
```
This conflicts a bit with `layer.WithMediaType`. I might just drop the `else` here? We could also warn or error or document this fact.

LFrobeen replied:
I removed the mediaType override for both zstd and gzip in order to be consistent. If `WithCompression(compression.ZStd)` is used, then the user should also set `WithMediaType(`types.OCILayerZStd)`. I've mentioned that in the updated documentation for `WithCompression`. When the layer is created via `LayerFromOpener`, the method will check whether the layer's `mediaType` matches its `compression` and print a warning if not. What do you think about this approach?

jonjohnsonjr commented:
There are a handful of submodules here (sorry) -- you might need to `go mod tidy` some others:

```
$ find | grep go\.mod | grep -v \.git
./cmd/krane/go.mod
./pkg/authn/k8schain/go.mod
./pkg/authn/kubernetes/go.mod
./go.mod
```

LFrobeen commented:
Oh yes, I totally overlooked those. I ran `go mod tidy` in all the submodules now. Indeed `github.com/klauspost/compress v1.15.11 // indirect` was missing from `pkg/authn/k8schain/go.mod`.

imjasonh commented:
Not annoying at all!

LFrobeen commented:
Awesome, thanks for triggering the tests! It looks like all checks have passed this time.

jonjohnsonjr commented:
Thanks for sticking with this -- I'd like to play with it a bit before cutting a release, so I might break the API in small ways, but it looks pretty close to what we want.
